internal/stats/metrics.go

```diff
@@ -1,29 +1,200 @@
 
+	Type MetricType `json:"type"`
+
+	Value float64 `json:"value"`
+
 	// Additional attributes that are not standard enough to be formalized in fields.
-	Attributes map[string]any `json:"attributes,omitempty"`
+	Attributes map[MetricAttribute]any `json:"attributes,omitempty"`
+}
+
+var knownMetricTypes = map[MetricType]bool{
+	UnknownMetricAttributeType: true,
+	APIRequestStatsType:        true,
+	CodeAnalysisTimeStatsType:  true,
+	GradingTimeStatsType:       true,
+	TaskTimeStatsType:          true,
 }
 
-func (this BaseMetric) GetTimestamp() timestamp.Timestamp {
+var knownMetricAttributes = map[MetricAttribute]bool{
+	UnknownMetricAttributeKey: true,
+	AssignmentIDKey:           true,
+	AnalysisTypeKey:           true,
+	CourseIDKey:               true,
+	EndpointKey:               true,
+	LocatorKey:                true,
+	SenderKey:                 true,
+	TaskTypeKey:               true,
+	UserEmailKey:              true,
+}
+
+func (this Metric) GetTimestamp() timestamp.Timestamp {
 	return this.Timestamp
 }
 
-type SystemMetrics struct {
-	BaseMetric
+func (this *Metric) LogValue() []*log.Attr {
+	attrs := []*log.Attr{
+		log.NewAttr("metric-type", this.Type),
+	}
+
+	for key, value := range this.Attributes {
+		attrs = append(attrs, log.NewAttr(string(key), value))
+	}
+
+	return attrs
+}
+
+func (this *Metric) Validate() error {
+	if this == nil {
+		return fmt.Errorf("No metric was given.")
+	}
+
+	err := validateType(this.Type)
+	if err != nil {
+		return err
+	}
+
+	if this.Timestamp.IsZero() {
+		return fmt.Errorf("Metric timestamp was not set.")
+	}
+
+	if this.Attributes == nil {
+		this.Attributes = make(map[MetricAttribute]any)
+	}
+
+	return validateAttributeMap(this.Attributes)
+}
+
+func validateAttributeMap(attributes map[MetricAttribute]any) error {
+	if attributes == nil {
+		return nil
+	}
+
+	for key, value := range attributes {
+		if key == UnknownMetricAttributeKey {
+			return fmt.Errorf("Attribute key was empty.")
+		}
+
+		if !knownMetricAttributes[key] {
+			return fmt.Errorf("Invalid attribute: '%v'.", key)
+		}
+
+		if value == nil || value == "" {
+			return fmt.Errorf("Attribute value was empty for key '%v'.", key)
+		}
+	}
+
+	return nil
+}
+
+func validateType(metricType MetricType) error {
+	if metricType == UnknownMetricAttributeType {
+		return fmt.Errorf("Type was not set.")
+	}
+
+	if !knownMetricTypes[metricType] {
+		return fmt.Errorf("Invalid metric type: '%v'.", metricType)
+	}
+
+	return nil
+}
+
+func AsyncStoreMetric(metric *Metric) {
+	err := metric.Validate()
+	if err != nil {
+		log.Error("Failed to validate metric.", err)
+		return
+	}
+
+	storeFunc := func() {
+		err := StoreMetric(metric)
+		if err != nil {
+			log.Error("Failed to store metric.", metric)
+			return
+		}
+	}
+
+	if config.UNIT_TESTING_MODE.Get() {
+		storeFunc()
+	} else {
+		go storeFunc()
+	}
+}
+
+func (this *Metric) setAttrIfNotEmpty(key MetricAttribute, value string) {
```

Comment:
s/Attr/Attribute/

Since the field uses the fill name, the method should as well.